src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseStudentParticipationRepository.java

```diff
@@ -164,6 +166,19 @@ List<ProgrammingExerciseStudentParticipation> findWithSubmissionsByExerciseIdAnd
     Optional<ProgrammingExerciseStudentParticipation> findWithSubmissionsByExerciseIdAndStudentLoginAndTestRun(@Param("exerciseId") long exerciseId,
             @Param("username") String username, @Param("testRun") boolean testRun);
 
+    @Query("""
+            SELECT participation.repositoryUri
+            FROM ProgrammingExerciseStudentParticipation participation
+                JOIN TREAT (participation.exercise AS ProgrammingExercise) pe
+            WHERE participation.repositoryUri IS NOT NULL
+                AND (
+                    (pe.dueDate IS NOT NULL AND :earliestDate <= pe.dueDate AND pe.dueDate <= :latestDate)
+                    OR (pe.exerciseGroup IS NOT NULL AND :earliestDate <= pe.exerciseGroup.exam.endDate AND pe.exerciseGroup.exam.endDate <= :latestDate)
```

Comment:
JPQL paths create implicit inner joins. Writing `pe.exerciseGroup.exam.endDate` anywhere in the query would result in something like `JOIN pe.exerciseGroup eg JOIN eg.exam e`. This means that if `exerciseGroup` is `NULL` the row is discarded before the `IS NOT NULL` check happens.
Maybe add some `LEFT JOIN`s to get the `exam`.